### ISSUE TITLE
Paper #508: Search-Based Adversarial Testing and Improvement of Constrained Credit Scoring Systems

### ISSUE THREAD

**Issue by kstolee** [researchart/fse20#75]
This issue is related to PR #51 

Authors are applying for the *reusable* badge.

Author contact:@yamizi

**Comment by kstolee** [comment]
# Authors

## Author 1 (Corresponding Author)

Salah Ghamizi

E-mail: <salah.ghamizi@uni.lu>

GitHub: [yamizi](https://github.com/yamizi)

## Author 2

Thibault Simonetto

E-mail: <thibault.simonetto.001@student.uni.lu>

GitHub: [thibaultsmnt](https://github.com/thibaultsmnt)

## Author 3

Maxime Cordy

E-mail: <maxime.cordy@uni.lu>

GitHub: [maxcordy](https://github.com/maxcordy)

**Comment by art-reviewer** [comment]
[Paper](https://github.com/yamizi/fse20/blob/19c532985a4989c0a7b7a708a8ed3865b7dc85fe/submissions/reusable/fse-508/PAPER-ACCEPTED.pdf)

**Comment by softwareArtifactsEvaluator** [comment]
_Applying for: **Reusable**_

Thank you for submitting the artifacts and congratulations on the paper!

Based on the current version of submission, I cannot recommend the reusable badge, however I would be happy to check the artifacts again and accordingly update my recommendation if the setup issues mentioned below can be resolved:

## Comments:

- Please also include simple instructions for installing conda (the few commands needed for installation)

- Does the zip file require a password because of the NDA? (I think with an open source license it might be better to remove the password, if possible).

-  I was not able to successfully extract the zip file using `zip` package, and `7zip` was required for extracting the file. This should either be fixed to support default `zip` utilities, or be included as a required dependency in the instructions:
    ```
    Archive:  coeva2.zip
    skipping: analysis.py             need PK compat. v6.3 (can do v4.6)
    skipping: configurations/config_1.json  need PK compat. v6.3 (can do v4.6)
    skipping: configurations/config_f1f2f3_fast.json  need PK compat. v6.3 (can do v4.6)
    skipping: configurations/config_f1f2_fast.json  need PK compat. v6.3 (can do v4.6)
    skipping: configurations/config_f1f3_fast.json  need PK compat. v6.3 (can do v4.6)
    skipping: configurations/papernot_fast.json  need PK compat. v6.3 (can do v4.6)
    skipping: configurations/random_fast.json  need PK compat. v6.3 (can do v4.6)
    skipping: custom.py               need PK compat. v6.3 (can do v4.6)
    ...
    ```

- `environments.yml` includes a hard-coded path specific to your machine:
    ```
    prefix: C:\Users\salah.ghamizi\AppData\Local\Continuum\anaconda3\envs\coeva2
    ```

- The current versions of dependencies specified in `environment.yml` are not found when conda is trying to create the environment:
    ```
    $ conda env create -f environment.yml
    Collecting package metadata (repodata.json): done
    Solving environment: failed

    ResolvePackageNotFound: 
    - openssl==1.1.1g=he774522_0
    - wincertstore==0.2=py37_0
    - vs2015_runtime==14.16.27012=hf0eaf9b_2
    - zlib==1.2.11=h62dcd97_4
    - python==3.7.7=h81c818b_4
    - vc==14.1=h0510ff6_4
    - sqlite==3.31.1=h2a8f88b_1
    ```

    ```
    $ conda --version
    conda 4.8.2
    ```

- Considering the difficulty of setting up environment, a packaged environment (such as a Docker or virtual machine image) which includes all the dependencies can significantly help with executing the artifacts in the future.

---

- :question: **Publicly available**: a DOI link is provided and an open source license is used. _However the zip file requires a password which is only available in the README.md in fse20 repository._ 
- :heavy_multiplication_x: **Documentation**: The scripts and directories are well documented however _I was not able to execute any of the scripts because of the current issues in setting up the environment._
- **Consistent/Complete**: The artifacts are relevant to the papers, however as the work is subject to NDA, the authors are not able to include the dataset that was used in the study.
- :heavy_multiplication_x: **Excercisable**: The repository includes scripts which can be used on another publicly available dataset, _However,I was not able to setup the environment to execute them._

**Comment by seartifact** [comment]
Hi @yamizi, 

Overall the documentation you provided goes into detail in terms of installation, set-up the environment, and running the tool. 

I tried installing the artifact on Ubuntu 18.10 with 16GB RAM.

## Available
The artifact is uploaded to Zenodo

## Functional
I had trouble setting up the artifact. The README is detailed, but I couldn't get it to run. 

I ran into the same challenges as described by @softwareArtifactsEvaluator 
+ unzip didn't work. Had to install 7z for it. 
+ Encountered ResolvePackageNotFound error. 

To summarize:
- The artifact is upload to Zenodo (so available). However, the dataset is under NDA--hence not included as a part of this study.
- However, I couldn't set up the environment to validate the functionality.

**Comment by yamizi** [comment]
Dear reviewers @softwareArtifactsEvaluator @seartifact 

Thank you for your valuable feedback. We react first to the installation issues and then to the individual questions.

***********************
* INSTALLATION ISSUES * 
***********************

As we understand it, the Conda installation from the environment fails on Ubuntu because Conda attempts to set up Windows-specific dependencies. We are sorry that you struggled with the installation because of this problem.

As a result, we have fixed this issue and prepared a virtual machine running on Linux that contains everything one needs to install our tool and run our experiments without issue (and without relying on Conda). Should you be willing to give our artefact a second try, we provide you with a zip archive containing the virtual machine. 

You can get the virtual machine here:
https://drive.google.com/file/d/1yJpUcWB9mg4GaFDAwJCtFLYGbLVwODmK/view?usp=sharing

It is a Vmware virtual machine that can be launched on Linux with Vmware player: 
https://my.vmware.com/en/web/vmware/downloads/details?downloadGroup=PLAYER-1556&productId=800&rPId=47861

The coeva directory is located in the "Documents" directory. From the coeva directory, you can run the various scripts presented in README.md (as provided in the original artefact submission). The root password is "fse".

If you do not want to use the virtual machine, you can download this compressed version (without a password and no 7zip): https://drive.google.com/file/d/1ahXfIro5ccCzAHfJG-vGzkjHMN8-4URU/view?usp=sharing
   - if you have conda you can run install.sh
   - otherwise, please use python3.7 and run "pip install -r ./requirements.txt" (or pip3).


************************
* INDIVIDUAL QUESTIONS *
************************

[R1] Please also include simple instructions for installing conda (the few commands needed for installation).

==> Thank you. We will update the repository with explanations on how to use Conda to get started.

[R1] Does the zip file require a password because of the NDA? (I think with an open source license it might be better to remove the password, if possible).

==> We had to wait for the partner's approval before releasing the tool. To submit the artefact, we had no other choice but to restrict the access to the reviewers (hence the password). Since then, we have received the approval and will release the source code as open source shortly.


[R1] I was not able to successfully extract the zip file using zip package, and 7zip was required for extracting the file. This should either be fixed to support default zip utilities, or be included as a required dependency in the instructions
[R2] unzip didn't work. Had to install 7z for it.

==> We used 7zip to increase the portability of the archive and optimize the compression. The archive could be unzipped with the default tool in MacOs and Windows. We are sorry that you had issues. Nevertheless, the public version will be in plain code in an open git repository.

[R1] environments.yml includes a hard-coded path specific to your machine:

==> Thank you for pointing this. Please note that this prefix is ignored for fresh installation and only used for reference. We removed it as you suggested.

[R1] The current versions of dependencies specified in environment.yml are not found when conda is trying to create the environment:
[R2] Encountered ResolvePackageNotFound error.

==> Please see above our detailed answer to installation issues. Please accept our sincere apologies.

**Comment by gousiosg** [comment]
@seartifact @softwareArtifactsEvaluator Can you please comment on the author's changes again? Do it as soon as possible, as we need to make decisions by Jul 1

**Comment by seartifact** [comment]
Hi, 

Thank you for the updated artifacts. I downloaded the artifact from this [link](https://drive.google.com/file/d/1ahXfIro5ccCzAHfJG-vGzkjHMN8-4URU/view?usp=sharing). 

## Functional
- I was able to set up the environment using the install.sh command which was provided by the authors.
- The artifact is well documented. 
- Please note the file (experiments_results.py) referred to README.md is not present in the artifact.
- nit: Could you mention how much time it takes to run the experiments. 


## Reusable
The quality of the documentation is good. Authors provide enough information to reproduce as well as create new experiments. The experiments take really long to run. I am currently running 
```
python ./experiments.py -x random
```
I will update the review once it terminates. 

@gousiosg Please note that the artifact (I used for the evaluation) is downloaded from a Google drive and currently not a part of this repo

**Comment by softwareArtifactsEvaluator** [comment]
Thank you for addressing the previous comments!

I was able to use the virtual machine image and try running the scripts. I think this addition will make things easier if there are issues with setting up the environment. 

Comments:
- As @seartifact mentioned `experiments_results.py` is not present in the artifacts (or inside the VM).
- I also agree that it would be helpful to know approximately how long the experiments take to run (and the specs of system that authors used).
- Inside the VM, `analysis.py` contains the content of README instead of the script?

Reminders for public artifact repository: 
- Uploading updated artifacts without password in the repository (similar to the artifacts from Google Drive link)
- Include VM image in artifacts

After addressing items mentioned above, I believe the artifact submission satisfies the requirements for the reusable badge.

**Comment by kstolee** [comment]
Applying ```functional``` for now. 

@yamizi are you able to address the requests by @softwareArtifactsEvaluator to achieve ```reusable```? Importantly, are the artifacts from the Google drive also in the zenodo DOI so we can apply the ```available``` badge as well?

**Comment by gousiosg** [comment]
@yamizi Please let us know today whether you will do the required fixes to get the `available` and `reusable` badges

**Comment by kstolee** [comment]
I just saw the comment by @seartifact:
> * The artifact is upload to Zenodo (so available). However, the dataset is under NDA--hence not included as a part of this study.

According to the Zenodo, 
> The password for the archive is only available to reviewers of the Artifact track of ESEC/FSE 2020

So, it seems like it was made available for review only (by necessity!). We'll go with ```functional``` and consider the evaluation done.

**Comment by kstolee** [comment]
Thanks everyone for your time and efforts!

**Comment by yamizi** [comment]
Thank you for the review and your feedback. Sorry I couldn't get back to you immediately (for medical reasons).

-The Zenodo DOI only contains the raw version, it would have changed if we updated the files.
-The analysis.py script is the one to display the results (instead of experiments_results.py)
-To run the experiments with the same parameters of the paper, it would indeed take longer on a virtual machine. The paper experiments were run within the company infrastructure. The artifact on the smaller dataset took about 24h to run on Nvidia Quadro P3200

Thank you again for your reviews!
